pkg/provisioner/ironic/ironic.go

```diff
@@ -892,6 +893,31 @@ func (p *ironicProvisioner) Adopt() (result provisioner.Result, err error) {
 		err = fmt.Errorf("Invalid state for adopt: %s",
 			ironicNode.ProvisionState)
 	case nodes.Manageable:
+		if p.host.Spec.Image == nil {
+			// Without an image, adoption will fail. We can stop and
+			// wait for the image to be set, but we need to do that in
+			// a way that ensures we don't enter the states where we
+			// try to manage the host power status. Setting our own
+			// error message here avoids exposing the error message
+			// from Ironic that talks about fields in Ironic with
+			// names the user may not recognize.
+			result.ErrorMessage = "Image details missing for externally provisioned server."
```

Comment:
Is this actually an error? Can't we consider getting to the Manageable state a success? It means we got past Verifying, which is the important thing.

Reply:
We can't go any further without the image. So unless we add a new state to the BMH state machine for hosts without images, we have to do something here to keep it from progressing and ending up stuck in a failure when adoption doesn't work.

Maybe we do need another state?

Reply:
We're in the ExternallyProvisioned state... what further stuff do we have to do? Getting to Manageable means we can control the power, doesn't it?

Reply:
No, we have to adopt the host (see line 921).

Reply:
I've just noticed that we allow the Host to freely switch between the ExternallyProvisioned and Ready states without deprovisioning it first. That's likely a bug.

We have to adopt the host if we are in the Provisioned state. And we might want to adopt the host _if_ we know the image so that Host can transition directly from ExternallyProvisioned->Provisioned without cleaning/inspection (although we don't support that today... we just go from ExternallyProvisioned->Ready without cleaning lol). But I'm not aware of any reason now or in the future that we would *have* to adopt an externally-provisioned host when we haven't been told what image (if any) is expected to be running on it.

Reply:
> I've just noticed that we allow the Host to freely switch between the ExternallyProvisioned and Ready states without deprovisioning it first. That's likely a bug.

Fun. Maybe? I could go either way.

> We have to adopt the host if we are in the Provisioned state. And we might want to adopt the host _if_ we know the image so that Host can transition directly from ExternallyProvisioned->Provisioned without cleaning/inspection (although we don't support that today... we just go from ExternallyProvisioned->Ready without cleaning lol). But I'm not aware of any reason now or in the future that we would _have_ to adopt an externally-provisioned host when we haven't been told what image (if any) is expected to be running on it.

According to @dtantsur or @juliakreger when I was writing that code, Ironic only monitors the power state of hosts that are adopted. Maybe I'm mis-remembering that, though? In any case, adoption today requires the image, and that's why this fix is preventing adoption until the image is set.

Reply:
> According to @dtantsur or @juliakreger when I was writing that code, Ironic only monitors the power state of hosts that are adopted.

Ah, I hadn't heard that. It's possible - adopting moves it to the active state (i.e. provisioned), so that shouldn't be required per se, but it may be we need to get to the available state before we could manage the power. If you don't have an image, the way to get there is via cleaning, which we don't want to do on an externally provisioned host (until it goes back to Ready!).

If that's the case, then I can see why we would treat this as an error, since it means we can't change the power state while externally provisioned.

It's tempting to just pass some bogus image, but there's no way to replace it with the real one if we get it later without dropping the Ironic DB.